### PR TITLE
delete superversions in BackgroundCallPurge

### PR DESCRIPTION
**Pull request by xieyilun**
I found that CleanupSuperVersion() may block Get() for 30ms+ （per MemTable is 256MB）.

Then I found "delete sv" in ~SuperVersion() takes the time.

The backtrace looks like this

DBImpl::GetImpl() -> DBImpl::ReturnAndCleanupSuperVersion() ->
DBImpl::CleanupSuperVersion() : delete sv; -> ~SuperVersion()

I think it's better to delete in a background thread,  please review it。

### CODE REVIEW

File: db/column_family.cc [hunk 1]
```diff
@@ -1118,22 +1116,19 @@ SuperVersion* ColumnFamilyData::GetThreadLocalSuperVersion(
...
     RecordTick(ioptions_.statistics, NUMBER_SUPERVERSION_ACQUIRES);
-    SuperVersion* sv_to_delete = nullptr;
 
     if (sv && sv->Unref()) {
       RecordTick(ioptions_.statistics, NUMBER_SUPERVERSION_CLEANUPS);
-      db_mutex->Lock();
+      db->mutex()->Lock();
       // NOTE: underlying resources held by superversion (sst files) might
       // not be released until the next background job.
-      sv->Cleanup();
```
Comment (siying) [inline db/column_family.cc]: It feels more natural to keep this line here than moving it to DB::AddSuperVersionsToFreeQueue().

File: db/column_family.cc [hunk 2]
```diff
@@ -1118,22 +1116,19 @@ SuperVersion* ColumnFamilyData::GetThreadLocalSuperVersion(
...
-      sv_to_delete = sv;
+      db->AddSuperVersionsToFreeQueue(sv);
+      db->SchedulePurge();
```
Comment (siying) [inline db/column_family.cc]: Right now we only do it with option avoid_unnecessary_blocking_io is true. Can we keep this pattern? We may not have the information inside class ColumnFamilyData, but I think ImmutableDBOptions is passed into the constructor, so we an add a member variable of ColumnFamilyData `bool avoid_unnecessary_blocking_io` to remember it.

**Comment by siying** [comment]
I realized that code comment of `avoid_unnecessary_blocking_io` needs to be updated to include this new scenario too. Unfortunately the option name isn't be accurate enough (the one we are avoiding here is not I/Os but large memory delocation), but I think we can live with the confusing name but update the comment and explain the scenario.

File: db/db_test2.cc [hunk 3]
```diff
@@ -4210,6 +4210,42 @@ TEST_F(DBTest2, SeekFileRangeDeleteTail) {
...
+  size_t value = options.write_buffer_manager->memory_usage();
+  ASSERT_GT(value, base_value);
+
+  db_->GetEnv()->SetBackgroundThreads(1, Env::Priority::HIGH);
+  test::SleepingBackgroundTask sleeping_task_after;
+  db_->GetEnv()->Schedule(&test::SleepingBackgroundTask::DoSleepTask,
+                          &sleeping_task_after, Env::Priority::HIGH);
+  delete iter;
+
+  Env::Default()->SleepForMicroseconds(100000);
```
Comment (siying) [inline db/db_test2.cc]: This is not a best practice. The good practice is to use waiting. I believe in this case a sync point can do the job.

Comment (riversand963) [reply db/db_test2.cc]: @xieyilun do you have plan to address this comment? Sleeping can potentially introduce flakiness to tests.

Comment (xieyilun) [reply db/db_test2.cc]: The Sleeping just to prove the memtable was deleted in the BackgroundPurge job, it does not affect correctness. @siying didn't understand it correctly before, we discussed in Wechat and Reach an agreement.
image
image

Comment (riversand963) [reply db/db_test2.cc]: Thanks @xieyilun for the context.

**Pull request closed without merging**